### ISSUE TITLE
CWG2804 Lookup for determining rewrite targets

### ISSUE THREAD

**Issue by jensmaurer** [cplusplus/papers#1688]
Some cases for synthesized equality operator candidates yield possibly surprising outcomes, owing to the details of the "rewrite target" rules.

CWG seeks the advice of EWG to possible improve or change these rules.

See CWG2804 for details.

**Comment by erichkeane** [comment]
EWG discussed [CWG2804](https://cplusplus.github.io/CWG/issues/2804.html) during the [Tuesday Morning session](https://wiki.edg.com/bin/view/Wg21kona2023/NotesEWGCWG2804) in Kona.  The following poll was taken.

CWG2804 "Lookup for determining rewrite targets", the intent is to check for the existence of a declaration that differs only in the operator name (ie, `bx` should work, `by` should be ill-formed).
| SF | F | N | A | SA |
|------|---|-----|---|------|
|  4   |  11   |   3 |  0   |  0     |

Result: Consensus

**Comment by jfbastien** [comment]
@jensmaurer does this also resolve https://github.com/cplusplus/papers/issues/1645 ?

**Comment by jensmaurer** [comment]
@jfbastien : Yes, but this for sure makes the example in #1645 ill-formed. But we intentionally made this example well-formed very recently.  So, EWG seems to have reversed direction as far as the example in #1645 is concerned.

**Comment by jensmaurer** [comment]
The EWG decision partially reverts [P2468R2](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2468r2.html) (The Equality Operator You Are Looking For), applied in July 2022. In particular, the line `bool c1 = B() == C();` from the example added thereby is no longer well-formed.

**Comment by erichkeane** [comment]
EWG discussed this and request that the authors of P2468 (#1127 @brevzin et-al) review the original CWG2804 as well as the break that was caused by the fix to it, and come back with a more all-encompassing solution/set of questions as an omnibus as a revision to P2468.

**Comment by brevzin** [comment]
Tagging @zygoloid and... is Cameron not on github? What's his handle?

**Comment by frederick-vs-ja** [comment]
> Tagging @zygoloid and... is Cameron not on github? What's his handle?

I believe @cdacamar is him.

**Comment by cdacamar** [comment]
Just so I understand, is the idea that the lookup for the corresponding `!=` should be constrained to the same _lexical_ scope?  

Note: the [programming model](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2468r2.html#programming-model) suggested in P2468R2 was that if the programmer writes a corresponding `operator!=` then the rewrite behavior for `operator==` is dropped.  From the perspective of the programming model, `bx` should compile as should `by` (since the hidden friend function would indeed correspond to the `operator==` as it is in the same scope as `operator!=).

As for the standard reference:

```c++
  struct B {
    bool operator==(const B&); // #2
  };
  struct C : B {
    C();
    C(B);
    bool operator!=(const B&); // #3
  };
  bool c1 = B() == C(); // OK, calls #2; reversed #2 is not a candidate because search for operator!= in C finds #3
```

If we consider the programming model again, the sample should still compile as the scope _S_ is `B` since the two declarations would still correspond.  I believe the wording can change however it needs to but the original programming model should be retained.

**Comment by jensmaurer** [comment]
My understanding of the `B`/`C` example from the standard in particular was that this was representative of code found by Microsoft in the wild (possibly in their own corporate codebase), and that the wording was thus specifically crafted to avoid breakage there.

**Comment by jensmaurer** [comment]
Beyond that, I'd like to remind all interested parties that this paper tracker is for administrative purposes only. Any technical discussion should happen elsewhere, e.g. on the reflectors.

**Comment by erichkeane** [comment]
EWG discussed this during the [Monday AM](https://wiki.edg.com/bin/view/Wg21tokyo2024/NotesEWGCWG2804) session in Tokyo.  The following Poll was taken:

CWG2804 EWG agrees that this is an issue worth considering, but would like to see a paper to propose a change.

| SF | F | N | A | SA |
|----|---|---|---|----|
|  1 | 14| 7 | 0 |  0 |

Result: Consensus

No paper author was identified.

**Comment by hanickadot** [comment]
EWG St. Louis: @usx95  volunteered to write a paper
